src/librustc_mir/transform/promote_consts.rs

```diff
@@ -0,0 +1,404 @@
+//! move analysis runs after promotion on broken MIR.
+
+use rustc::mir::repr::*;
+use rustc::mir::visit::{LvalueContext, MutVisitor, Visitor};
+use rustc::ty::{self, TyCtxt};
+use syntax::codemap::Span;
+
+use build::Location;
+use traversal::ReversePostorder;
+
+use std::mem;
+
+/// State of a temporary during collection and promotion.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum TempState {
+    /// No references to this temp.
+    Undefined,
+    /// One direct assignment and any number of direct uses.
+    /// A borrow of this temp is promotable if the assigned
+    /// value is qualified as constant.
+    Defined {
+        location: Location,
+        uses: usize
+    },
+    /// Any other combination of assignments/uses.
+    Unpromotable,
+    /// This temp was part of an rvalue which got extracted
+    /// during promotion and needs cleanup.
+    PromotedOut
+}
+
+impl TempState {
+    pub fn is_promotable(&self) -> bool {
+        if let TempState::Defined { uses, .. } = *self {
+            uses > 0
+        } else {
+            false
+        }
+    }
+}
+
+/// A "root candidate" for promotion, which will become the
+/// returned value in a promoted MIR, unless it's a subset
+/// of a larger candidate.
+pub enum Candidate {
+    /// Borrow of a constant temporary.
+    Ref(Location),
+
+    /// Array of indices found in the third argument of
+    /// a call to one of the simd_shuffleN intrinsics.
+    ShuffleIndices(BasicBlock)
+}
+
+struct TempCollector {
+    temps: Vec<TempState>,
+    location: Location
+}
+
+impl<'tcx> Visitor<'tcx> for TempCollector {
+    fn visit_lvalue(&mut self, lvalue: &Lvalue<'tcx>, context: LvalueContext) {
+        self.super_lvalue(lvalue, context);
+        if let Lvalue::Temp(index) = *lvalue {
+            // Ignore drops, if the temp gets promoted,
+            // then it's constant and thus drop is noop.
+            if let LvalueContext::Drop = context {
+                return;
+            }
+
+            let temp = &mut self.temps[index as usize];
+            if *temp == TempState::Undefined {
+                match context {
+                    LvalueContext::Store |
+                    LvalueContext::Call => {
+                        *temp = TempState::Defined {
+                            location: self.location,
+                            uses: 0
+                        };
+                        return;
+                    }
+                    _ => {}
+                }
+            } else if let TempState::Defined { ref mut uses, .. } = *temp {
+                match context {
+                    LvalueContext::Borrow {..} |
+                    LvalueContext::Consume |
+                    LvalueContext::Inspect => {
+                        *uses += 1;
+                        return;
+                    }
+                    _ => {}
+                }
+            }
+            *temp = TempState::Unpromotable;
```

Comment:
Ah, perhaps disregard my prior two comments -- but add to the `{}` cases above something like:

```
{ /* mark as unpromotable below */ }
```